fgpyo/sam/__init__.py

```
@@ -1036,55 +1040,96 @@ class ReadEditInfo:
+                ins_bases += elem.length
+        elif op == CigarOp.D:  # consumes ref
+            if target_offset + elem.length >= len(reference_sequence):
+                break  # out of bounds
             deletions += 1
             del_bases += elem.length
-        elif op == CigarOp.M or op == CigarOp.X or op == CigarOp.EQ:
-            for i in range(0, elem.length):
-                q = rec.query_sequence[query_offset + i].upper()
-                t = reference_sequence[target_offset + i].upper()
-                if q != t or q not in ok_bases:
+            md_edits.append(str(current_match_count))  # append match count and reset
+            current_match_count = 0
+            md_edits.append(f"^{reference_sequence[target_offset:target_offset + elem.length]}")
+            target_offset += elem.length
+        elif op in (CigarOp.M, CigarOp.X, CigarOp.EQ):
+            for in_block_offset in range(0, elem.length):
+                if (target_offset + in_block_offset) >= len(reference_sequence):
+                    break  # out of bounds
+                query_base: str = rec.query_sequence[query_offset + in_block_offset].upper()
+                ref_base: str = reference_sequence[target_offset + in_block_offset].upper()
+
+                if query_base != ref_base or (query_base not in ok_bases and not n_as_match):
                     mms += 1
-                else:
+                    md_edits.append(str(current_match_count))
+                    current_match_count = 0
+                    md_edits.append(ref_base)  # grab mismatched base from the reference
+                else:  # match
                     matches += 1
+                    current_match_count += 1
+            query_offset += elem.length_on_query
+            target_offset += elem.length_on_target
+        elif op not in (CigarOp.H, CigarOp.P):  # if we come across an unexpected CIGAR operation
+            raise ValueError(f"Invalid CIGAR operation: {op}")
```

Issue: I don't think these are errors.

Both hard clipping and padding operators consume neither reference bases nor read bases (see [the table on page 8 in the spec](https://samtools.github.io/hts-specs/SAMv1.pdf)), so they can be silently ignored.  This is what the `htsjdk`  implementation does, which we're following.

Can you add a test cases with cigars that have hard clips and padding operators (separately)?


```suggestion
```

---

I am so sorry, I read this as `elif op in (CigarOp.H, CigarOp.P)`. Consider `assert_never`.  Sorry again

---

I could not get `assert_never()` to work here. I reviewed it with Clint and it sounded like a potential rabbit hole?
```
elif op not in (CigarOp.H, CigarOp.P):  # pragma: no cover
  assert_never(op)

error: Argument 1 to "assert_never" has incompatible type "Literal[CigarOp.M, CigarOp.I, CigarOp.S, CigarOp.H, CigarOp.P, CigarOp.EQ, CigarOp.X, CigarOp.code, CigarOp.character, CigarOp.consumes_query, CigarOp.consumes_reference]"; expected "Never"  [arg-type]
```